From 43e2e1220426fa7e0b740eed226de6e30f010e3e Mon Sep 17 00:00:00 2001
From: Thomas Lukaszewicz <tluk@chromium.org>
Date: Wed, 15 Jan 2025 13:01:21 -0800
Subject: [PATCH] [M132 Merge][discard] Ensure tabs remain eligible for successive discard operations

This CL updates PageDiscardingHelper to reset DiscardAttemptMarkers
for nodes following a main frame navigation.

This is necessary to ensure that a given tab remains eligible for
discarding in the specific case where:
  1. A discard attempt was previously made on the tab
  2. The tab was later navigated / reloaded and its content is
     refreshed

The current behavior assumes that the DiscardAttemptMarker is
destroyed following a discard operation, as the target WebContents
/ PageNode is replaced.

However with the new implementation the discarded tab's WebContents
is retained. This makes the tab inelibile for any future discard
operations, even after it has been re-activated  / reloaded.

See bug for additional context.

(cherry picked from commit 784691323d7924f76ec14f46cc4c2cc19337221b)

Bug: 386801193, 388096470
Change-Id: Id675d8a349d95d7f4ca34620c5ba57d254634210
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/6132978
Code-Coverage: findit-for-me@appspot.gserviceaccount.com <findit-for-me@appspot.gserviceaccount.com>
Reviewed-by: Francois Pierre Doray <fdoray@chromium.org>
Commit-Queue: Thomas Lukaszewicz <tluk@chromium.org>
Cr-Original-Commit-Position: refs/heads/main@{#1402667}
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/6148052
Cr-Commit-Position: refs/branch-heads/6834@{#3791}
Cr-Branched-From: 47a3549fac11ee8cb7be6606001ede605b302b9f-refs/heads/main@{#1381561}
---

--- a/chrome/browser/performance_manager/policies/page_discarding_helper.cc
+++ b/chrome/browser/performance_manager/policies/page_discarding_helper.cc
@@ -310,6 +310,7 @@
 
 void PageDiscardingHelper::OnPassedToGraph(Graph* graph) {
   DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+  graph->AddPageNodeObserver(this);
   graph->GetNodeDataDescriberRegistry()->RegisterDescriber(this,
                                                            kDescriberName);
 }
@@ -317,6 +318,7 @@
 void PageDiscardingHelper::OnTakenFromGraph(Graph* graph) {
   DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
   graph->GetNodeDataDescriberRegistry()->UnregisterDescriber(this);
+  graph->RemovePageNodeObserver(this);
 }
 
 const PageLiveStateDecorator::Data*
@@ -504,6 +506,14 @@
   return !it->second->MatchURL(url).empty();
 }
 
+void PageDiscardingHelper::OnMainFrameDocumentChanged(
+    const PageNode* page_node) {
+  // When activated a discarded tab will re-navigate, instantiating a new
+  // document. Ensure the DiscardAttemptMarker is cleared in these cases to
+  // ensure a given tab remains eligible for discarding.
+  DiscardAttemptMarker::Destroy(PageNodeImpl::FromNode(page_node));
+}
+
 base::Value::Dict PageDiscardingHelper::DescribePageNodeData(
     const PageNode* node) const {
   auto can_discard = [this, node](DiscardReason discard_reason) {
--- a/chrome/browser/performance_manager/policies/page_discarding_helper.h
+++ b/chrome/browser/performance_manager/policies/page_discarding_helper.h
@@ -113,7 +113,8 @@
 // PageDiscardingHelper::GetFromGraph(graph()).
 class PageDiscardingHelper
     : public GraphOwnedAndRegistered<PageDiscardingHelper>,
-      public NodeDataDescriberDefaultImpl {
+      public NodeDataDescriberDefaultImpl,
+      public PageNode::ObserverDefaultImpl {
  public:
   // Export discard reason in the public interface.
   using DiscardReason = ::mojom::LifecycleUnitDiscardReason;
@@ -177,6 +178,9 @@
   static void AddDiscardAttemptMarkerForTesting(PageNode* page_node);
   static void RemovesDiscardAttemptMarkerForTesting(PageNode* page_node);
 
+  // PageNodeObserver:
+  void OnMainFrameDocumentChanged(const PageNode* page_node) override;
+
  protected:
   void OnPassedToGraph(Graph* graph) override;
   void OnTakenFromGraph(Graph* graph) override;
--- a/chrome/browser/performance_manager/policies/page_discarding_helper_browsertest.cc
+++ b/chrome/browser/performance_manager/policies/page_discarding_helper_browsertest.cc
@@ -19,6 +19,7 @@
 #include "content/public/browser/page_navigator.h"
 #include "content/public/browser/web_contents.h"
 #include "content/public/browser/web_contents_observer.h"
+#include "content/public/common/content_features.h"
 #include "content/public/common/referrer.h"
 #include "content/public/test/browser_test.h"
 #include "content/public/test/browser_test_utils.h"
@@ -57,8 +58,16 @@
   base::RunLoop run_loop_;
 };
 
-class PageDiscardingHelperBrowserTest : public InProcessBrowserTest {
+class PageDiscardingHelperBrowserTest
+    : public InProcessBrowserTest,
+      public ::testing::WithParamInterface<bool> {
  protected:
+  PageDiscardingHelperBrowserTest() {
+    scoped_feature_list_.InitWithFeatureState(::features::kWebContentsDiscard,
+                                              GetParam());
+  }
+  ~PageDiscardingHelperBrowserTest() override = default;
+
   void SetUpOnMainThread() override {
     InProcessBrowserTest::SetUpOnMainThread();
     ASSERT_TRUE(embedded_test_server()->Start());
@@ -107,6 +116,11 @@
     favicon_watcher.Wait();
   }
 
+  base::WeakPtr<PageNode> GetPageNodeAtIndex(int index) {
+    return PerformanceManager::GetPrimaryPageNodeForWebContents(
+        browser()->tab_strip_model()->GetWebContentsAt(index));
+  }
+
   void ExpectImmediateDiscard(
       int index,
       DiscardReason discard_reason,
@@ -158,9 +172,12 @@
         browser()->tab_strip_model()->GetWebContentsAt(index)->WasDiscarded(),
         expected_result);
   }
+
+ private:
+  base::test::ScopedFeatureList scoped_feature_list_;
 };
 
-IN_PROC_BROWSER_TEST_F(PageDiscardingHelperBrowserTest, DiscardSpecificPage) {
+IN_PROC_BROWSER_TEST_P(PageDiscardingHelperBrowserTest, DiscardSpecificPage) {
   // Test urgent and proactive discards in a loop to avoid the overhead of
   // starting a new browser every time.
   // TODO(crbug.com/40899366): Add tests for all the other heuristics in
@@ -240,6 +257,75 @@
   }
 }
 
+// Regression test for crbug.com/386801193. Ensure discarded tabs remain
+// eligible for successive discard operations following a reactivation / reload.
+IN_PROC_BROWSER_TEST_P(PageDiscardingHelperBrowserTest,
+                       DiscardedTabEligibleForSuccessiveDiscards) {
+  // Add a new background tab.
+  OpenNewBackgroundPage();
+  EXPECT_EQ(browser()->tab_strip_model()->count(), 2);
+
+  tabs::TabInterface* tab1 = browser()->tab_strip_model()->GetTabAtIndex(0);
+  tabs::TabInterface* tab2 = browser()->tab_strip_model()->GetTabAtIndex(1);
+  EXPECT_EQ(browser()->tab_strip_model()->GetActiveTab(), tab1);
+
+  // Attempt to discard the background tab.
+  const auto attempt_discard = [this]() {
+    base::WeakPtr<PageNode> discard_target_page_node = GetPageNodeAtIndex(1);
+    base::RunLoop run_loop;
+    PerformanceManager::CallOnGraph(
+        FROM_HERE, base::BindLambdaForTesting([&](Graph* graph) {
+          ASSERT_TRUE(discard_target_page_node);
+          auto* helper = PageDiscardingHelper::GetFromGraph(graph);
+          ASSERT_TRUE(helper);
+          EXPECT_EQ(
+              CanDiscardResult::kEligible,
+              helper->CanDiscard(discard_target_page_node.get(),
+                                 DiscardReason::URGENT, base::TimeDelta()));
+          helper->DiscardAPage(
+              base::BindLambdaForTesting(
+                  [&](std::optional<base::TimeTicks> first_discarded_at) {
+                    EXPECT_TRUE(first_discarded_at.has_value());
+                    run_loop.Quit();
+                  }),
+              DiscardReason::URGENT, base::TimeDelta());
+        }));
+    run_loop.Run();
+  };
+  attempt_discard();
+
+  // Ensure the background tab has been discarded.
+  EXPECT_FALSE(tab1->GetContents()->WasDiscarded());
+  EXPECT_TRUE(tab2->GetContents()->WasDiscarded());
+
+  // Activate and reload the discarded background page.
+  content::TestNavigationObserver reload_waiter(tab2->GetContents(), 1);
+  browser()->tab_strip_model()->ActivateTabAt(1);
+  EXPECT_EQ(browser()->tab_strip_model()->GetActiveTab(), tab2);
+  reload_waiter.Wait();
+
+  EXPECT_FALSE(tab1->GetContents()->WasDiscarded());
+  EXPECT_FALSE(tab2->GetContents()->WasDiscarded());
+
+  // Background the discarded tab again and attempt another discard.
+  browser()->tab_strip_model()->ActivateTabAt(0);
+  EXPECT_EQ(browser()->tab_strip_model()->GetActiveTab(), tab1);
+  attempt_discard();
+
+  // Ensure the background tab has been discarded again.
+  EXPECT_FALSE(tab1->GetContents()->WasDiscarded());
+  EXPECT_TRUE(tab2->GetContents()->WasDiscarded());
+}
+
+INSTANTIATE_TEST_SUITE_P(
+    ,
+    PageDiscardingHelperBrowserTest,
+    ::testing::Values(false, true),
+    [](const ::testing::TestParamInfo<
+        PageDiscardingHelperBrowserTest::ParamType>& info) {
+      return info.param ? "RetainedWebContents" : "UnretainedWebContents";
+    });
+
 }  // namespace
 
 }  // namespace performance_manager::policies
